.github/workflows/vm-kernel.yaml

```diff
@@ -28,39 +33,97 @@ on:
         type: string
         required: false
         default: ''
+      force-rebuild:
+        description: 'Rebuild the kernel image even if it already exists. No-op if `kernel-image-tag-override` is set'
+        type: boolean
+        required: false
+        default: false
     outputs:
       image:
         description: 'vm-kernel Docker image'
-        value: ${{ jobs.check-kernel-image-override.outputs.image || jobs.vm-kernel.outputs.image }}
+        value: ${{ jobs.check-kernel-image.outputs.image || jobs.vm-kernel.outputs.image }}
 
 env:
   VM_KERNEL_IMAGE: "neondatabase/vm-kernel"
 
 jobs:
-  check-kernel-image-override:
+  check-kernel-image:
     outputs:
-      image: ${{ steps.check-kernel-image-override.outputs.image }}
+      image: ${{ steps.get-kernel-image.outputs.image }}
+      last-kernel-sha: ${{ steps.get-last-kernel-commit-sha.outputs.last-kernel-sha }}
 
     runs-on: ubuntu-latest
 
     steps:
-      - id: check-kernel-image-override
+      - name: get last kernel commit sha
+        id: get-last-kernel-commit-sha
+        env:
+          COMMIT_SHA: ${{ github.event.pull_request.head.sha || github.sha }}
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          CACHE_TAG=$(
+            gh api \
+              -H "Accept: application/vnd.github+json" \
+              -H "X-GitHub-Api-Version: 2022-11-28" \
+              --method GET \
+              --field path=neonvm/hack/kernel \
+              --field sha=${COMMIT_SHA} \
+              --field per_page=1 \
+              --jq ".[0].sha" \
+              "/repos/${GITHUB_REPOSITORY}/commits"
+          )
+          echo "last-kernel-sha=${CACHE_TAG}" >> $GITHUB_OUTPUT
+
+      - name: get kernel image
+        id: get-kernel-image
         env:
           OVERRIDE_TAG: ${{ inputs.kernel-image-tag-override }}
+          FORCE_REBUILD: ${{ inputs.force-rebuild }}
+          CACHE_TAG: ${{ steps.get-last-kernel-commit-sha.outputs.last-kernel-sha }}
         run: |
           if [ -n "${OVERRIDE_TAG}" ]; then
             DIGEST=$(docker manifest inspect ${VM_KERNEL_IMAGE}:${OVERRIDE_TAG} -v | jq -r '.Descriptor.digest')
             IMAGE="${VM_KERNEL_IMAGE}:${OVERRIDE_TAG}@${DIGEST}"
+          elif [ "${FORCE_REBUILD}" == "false" ]; then
+            CACHE_TAG_DIGEST=$(docker manifest inspect ${VM_KERNEL_IMAGE}:${CACHE_TAG} -v | jq -r '.Descriptor.digest' || true)
+            if [ -n "${CACHE_TAG_DIGEST}" ]; then
+              IMAGE="${VM_KERNEL_IMAGE}:${CACHE_TAG}@${CACHE_TAG_DIGEST}"
+            else
+              IMAGE=""
+            fi
+          elif [ "${FORCE_REBUILD}" == "true" ]; then
+            IMAGE=""
           else
             IMAGE=""
           fi
+
           echo "image=${IMAGE}" >> $GITHUB_OUTPUT
 
+      - name: login to docker hub
+        if: |
+          steps.get-kernel-image.outputs.image != '' &&
+          inputs.tag && !inputs.force-rebuild && !inputs.kernel-image-tag-override
```

Comment:
Took a while for me to validate that this is correct - would be good to have a comment here explaining that we only need to login if

1. there's a preexisting image to use (`steps.get-kernel-image.outputs.image != ''`), and we don't need to rebuild it (`!inputs.force-rebuild`); and
2. we explicitly want to tag the image (`inputs.tag != ''`)

(didn't include `kernel-image-tag-override` because I'm not sure I understand the semantics)

Reply:
Moved this to a separate step with comments. I also renamed a bunch of stuff to make it clear. Let me know what do you think

Reply:
Looks good :)